### ISSUE TITLE
Cannot use multiple wkhtmltox.dll file instances simultaneously in a process

### ISSUE THREAD

**Issue by tuespetre** [tuespetre/TuesPechkin#63]
Example:

```
        private static IConverter converter =
            new ThreadSafeConverter(
                new RemotingToolset<PdfToolset>(
                    someDeployment));

        private static IConverter anotherConverter =
            new ThreadSafeConverter(
                new RemotingToolset<PdfToolset>(
                    someOtherDeployment));
```

If both of these converters are used simultaneously, one of them is going to just hang at 10% when it tries to convert.

**Comment by tuespetre** [comment]
http://stackoverflow.com/questions/1202744/dynamically-p-invoking-a-dll

**Comment by tuespetre** [comment]
Explanation, since all I did was post a link:

When using the `[DllImport]` attribute, P/Invoke (part of .NET framework) will only resolve the location of the targeted DLL one time. So in order to make this work, new versions of the `PdfAssembly` and `ImageAssembly` classes need to be created that, instead of relying on `[DllImport]`, instead use `GetProcAddress` and a number of other WinAPI calls to properly hook into a specified copy of wkhtmltox.dll.

**Comment by lkm** [comment]
@tuespetre Have you made any progress on this? We're looking to use it in production but performance-wise it's much slower than the process.start approach of integrating wkhtmltopdf.

**Comment by jesusmogollon** [comment]
@tuespetre, I'm using it already and my only problem is when I release a new update to my website the previous converter is still on memory and in the next call is going to hang.

I figure this is related to the same memory allocation issue.

My only work around is to stop the website and the application pool, do the update and start everything back, if I don't do that it is a big problem to manually find and kill the website process.

I'm eager to see if you have made any progress on this.

Thanks a lot

**Comment by Rumbers** [comment]
In case this helps anyone... I had this same hanging behaviour when two applications using TuesPechkin singletons were hosted in IIS on the same app pool, and managed to get it working by giving each of the applications their own pool. Changing the StaticDeployment path didn't seem to help.

**Comment by basicvn** [comment]
I have same situation when building website that allow generate pdf for product offers. I also use ImageToolset to generate preview for PDF on site. App crash after few runs, and sometime hang.

I put the Image preview on separate WCF service then everything ok now. Beside of that, I also call PdfToolset.Unload to release the memory when multiple user try using the tool at same time.

The wkhtmltox.dll is now 0.12.4, I just updated the dll into x64 project and the Unload issue also gone (with wkhtmltox 0.12.1 each time I called Unload of PdfToolset before, the website crash after 2-3 minutes with a note in event viewer that memory violence).
